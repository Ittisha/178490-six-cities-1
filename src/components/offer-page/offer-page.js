import * as React from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';

import {getNearestOffers, getOffer} from '../../reducers/selectors';
import offerPropTypes from '../../props/apartment';
import {addPluralS} from '../../utils/addPluralS';
import {APARTMENT_TYPES} from '../../consts';
import {getRatingPercent} from '../../utils/get-rating-percent';
import ReviewSection from '../review-section/review-section';
import {CitiesMap} from '../cities-map/cities-map';
import {CardList} from '../card-list/card-list';

const MAX_PHOTO_NUMBER = 6;

export const OfferPage = ({offer, nearestOffers}) => {
  const {
    images,
    title,
    id,
    isPremium,
    type,
    bedrooms,
    maxAdults,
    price,
    goods,
    host,
    description,
    rating,
    cityName,
    cityCoords,
    cityZoom,
  } = offer;

  const city = {
    name: cityName,
    coords: cityCoords,
    zoom: cityZoom,
  };

  const apartmentsCoords = [...nearestOffers, offer].map(({id: idCurrent, coordinates, zoom}) => ({id: idCurrent, coordinates, zoom}));

  const wordBedroomForm = addPluralS(bedrooms, `Bedroom`);
  const wordAdultForm = addPluralS(maxAdults, `adult`);

  const premiumPropertyMark = isPremium
    ? (<div className="property__mark"><span>Premium</span></div>)
    : null;
  const proHostAvatarClass = host.isPro ? `property__avatar-wrapper--pro` : `;`;
  const proHostStatus = host.isPro ? <span className="property__user-status">Pro</span> : null;

  return (
    <React.Fragment>
      <main className="page__main page__main--property">
        <section className="property">
          <div className="property__gallery-container container">
            <div className="property__gallery">
              {images.slice(0, MAX_PHOTO_NUMBER).map((image, i) => <div
                key={`image-${id}-${i}`}
                className="property__image-wrapper"
              >
                <img className="property__image" src={image} alt={title} />
              </div>)}
            </div>
          </div>
          <div className="property__container container">
            <div className="property__wrapper">
              {premiumPropertyMark}
              <div className="property__name-wrapper">
                <h1 className="property__name">
                  {title}
                </h1>
                <button className="property__bookmark-button button" type="button">
                  <svg className="property__bookmark-icon" width="31" height="33">
                    <use xlinkHref="#icon-bookmark"></use>
                  </svg>
                  <span className="visually-hidden">To bookmarks</span>
                </button>
              </div>
              <div className="property__rating rating">
                <div className="property__stars rating__stars">
                  <span style={{width: `${getRatingPercent(rating)}`}}></span>
                  <span className="visually-hidden">Rating</span>
                </div>
                <span className="property__rating-value rating__value">{rating}</span>
              </div>
              <ul className="property__features">
                <li className="property__feature property__feature--entire">
                  {APARTMENT_TYPES[type]}
                </li>
                <li className="property__feature property__feature--bedrooms">
                  {`${bedrooms} ${wordBedroomForm}`}
                </li>
                <li className="property__feature property__feature--adults">
                  Max {`${maxAdults} ${wordAdultForm}`}
                </li>
              </ul>
              <div className="property__price">
                <b className="property__price-value">&euro;{price}</b>
                <span className="property__price-text">&nbsp;night</span>
              </div>
              <div className="property__inside">
                <h2 className="property__inside-title">What&apos;s inside</h2>
                <ul className="property__inside-list">
                  {goods.map((item, i) => <li
                    key={`good-${id}-${i}`}
                    className="property__inside-item"
                  >
                    {item}
                  </li>)}
                </ul>
              </div>
              <div className="property__host">
                <h2 className="property__host-title">Meet the host</h2>
                <div className="property__host-user user">
                  <div className={`property__avatar-wrapper user__avatar-wrapper ${proHostAvatarClass}`}>
                    <img
                      className="property__avatar user__avatar"
                      src={`/${host.avatar}`}
                      width="74"
                      height="74"
                      alt="Host avatar"
                    />
                  </div>
                  <span className="property__user-name">
                    {host.name}
                  </span>
                  {proHostStatus}
                </div>
                <div className="property__description">
                  <p className="property__text">
                    {description}
                  </p>
                </div>
              </div>
              <ReviewSection offerId={id} />
            </div>
          </div>
          <section className="property__map map">
            <CitiesMap
              apartmentsCoords={apartmentsCoords}
              city={city}
              activeItem={offer}
            />
          </section>
        </section>
        <div className="container">
          <section className="near-places places">
            <h2 className="near-places__title">Other places in the neighbourhood</h2>
            <CardList
              apartments={nearestOffers}
              cardListClass={`near-places__list`}
              cardClass={`near-places__card`}
            />
          </section>
        </div>
      </main>
    </React.Fragment>
  );
};

OfferPage.propTypes = {
  offer: offerPropTypes,
  nearestOffers: PropTypes.arrayOf(offerPropTypes),
};

const mapStateToProps = (state, ownProps) => {
  const currentOfferId = ownProps.match.params.id;
  const getNearestSelector = getNearestOffers(currentOfferId, 3);
  return Object.assign({}, ownProps, {
    offer: getOffer(state, currentOfferId),
    nearestOffers: getNearestSelector(state),
  });
};

export default connect(mapStateToProps)(OfferPage);
