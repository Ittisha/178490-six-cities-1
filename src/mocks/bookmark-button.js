export const MOCK_PROPS = {
  offerId: 1,
  isInBookmarks: false,
  className: ``,
  width: 15,
  height: 12,
};
