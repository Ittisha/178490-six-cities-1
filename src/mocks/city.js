export const INIT_CITY = {
  name: `Amsterdam`,
  coords: [52.38333, 4.9],
  zoom: 12,
};

export const MOCK_CITIES = [`Amsterdam`, `Paris`];

export const MOCK_CITIES_COORDS = {
  Amsterdam: [52.38333, 4.9],
  Paris: [48.864716, 2.349014],
};

export const MOCK_CITIES_ZOOM = {
  Amsterdam: 17,
  Paris: 15,
};
