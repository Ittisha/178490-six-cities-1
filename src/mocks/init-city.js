export const INIT_CITY = {
  name: `Amsterdam`,
  coords: [52.38333, 4.9],
};
