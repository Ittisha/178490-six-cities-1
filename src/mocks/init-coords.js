export const INIT_COORDS = [52.38333, 4.9];
