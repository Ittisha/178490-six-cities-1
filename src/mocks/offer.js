export const OFFER = {
  "city": {
    "name": `Cologne`,
    "location": {
      "latitude": 50.938361,
      "longitude": 6.959974,
      "zoom": 13}},
  "preview_image": `https://es31-server.appspot.com/six-cities/static/hotel/17.jpg`,
  "images": [
    `https://es31-server.appspot.com/six-cities/static/hotel/20.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/2.jpg`,
  ],
  "title": `The house among olive `,
  "is_favorite": false,
  "is_premium": false,
  "rating": 2.5,
  "type": `house`,
  "bedrooms": 2,
  "max_adults": 6,
  "price": 757,
  "goods": [
    `Coffee machine`,
    `Baby seat`,
    `Dishwasher`,
    `Fridge`,
    `Washing machine`,
    `Breakfast`,
  ],
  "host": {
    "id": 25,
    "name": `Angelina`,
    "is_pro": true,
    "avatar_url": `img/avatar-angelina.jpg`
  },
  "description": `This is a place for dreamers to reset, reflect, and create. Designed with a 'slow' pace in mind, our hope is that you enjoy every part of your stay; from making local coffee by drip in the morning, choosing the perfect record to put on as the sun sets.`,
  "location": {
    "latitude": 50.945361,
    "longitude": 6.935974,
    "zoom": 16
  },
  "id": 1
};
