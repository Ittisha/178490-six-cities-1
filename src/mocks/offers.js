export const OFFERS = [
  {
    isPremium: true,
    isInBookmarks: false,
    title: `Beautiful & luxurious apartment at great location`,
    photoUrl: `img/apartment-01.jpg`,
    price: 120,
    type: `Apartment`,
    rating: `93%`,
    id: 1,
    coordinates: [52.3909553943508, 4.85309666406198],
  },
  {
    isPremium: false,
    isInBookmarks: true,
    title: `Wood and stone place`,
    photoUrl: `img/room.jpg`,
    price: 80,
    type: `Private room`,
    rating: `80%`,
    id: 2,
    coordinates: [52.369553943508, 4.85309666406198],
  },
  {
    isPremium: false,
    isInBookmarks: false,
    title: `Canal View Prinsengracht`,
    photoUrl: `img/apartment-02.jpg`,
    price: 132,
    type: `Apartment`,
    rating: `80%`,
    id: 3,
    coordinates: [52.3909553943508, 4.929309666406198],
  },
  {
    isPremium: false,
    isInBookmarks: false,
    title: `Nice, cozy, warm big bed apartment`,
    photoUrl: `img/apartment-03.jpg`,
    price: 180,
    type: `Apartment`,
    rating: `80%`,
    id: 4,
    coordinates: [52.3809553943508, 4.939309666406198],
  }
];
