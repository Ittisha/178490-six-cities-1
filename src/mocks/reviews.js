export const MOCK_REVIEWS = [
  {
    id: 1,
    user: {
      avatarUrl: ``,
      id: 1,
      isPro: false,
      name: ``,
    },
    rating: 4,
    comment: ``,
    date: ``,
  },
  {
    id: 2,
    user: {
      avatarUrl: ``,
      id: 5,
      isPro: false,
      name: ``,
    },
    rating: 3,
    comment: ``,
    date: ``,
  }
];

export const MOCK_REVIEW = {
  comment: ``,
  id: 1,
  rating: `3`,
};
