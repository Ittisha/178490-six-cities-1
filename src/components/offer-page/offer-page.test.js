import * as React from 'react';
import {OfferPage} from './offer-page';
import {OFFER_MAPPED} from '../../mocks/offer';
import {OFFERS} from '../../mocks/offers';
import ShallowRenderer from 'react-test-renderer/shallow';

it(`OfferPage component is rendered properly`, () => {
  const renderer = new ShallowRenderer();
  const tree = renderer.render(
      <OfferPage
        offer={OFFER_MAPPED}
        nearestOffers={[OFFERS[3]]}
      />);

  expect(tree).toMatchSnapshot();
});
