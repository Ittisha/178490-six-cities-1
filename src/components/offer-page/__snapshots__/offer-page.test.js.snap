// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OfferPage component is rendered properly 1`] = `
<React.Fragment>
  <main
    className="page__main page__main--property"
  >
    <section
      className="property"
    >
      <div
        className="property__gallery-container container"
      >
        <div
          className="property__gallery"
        >
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Beautiful & luxurious apartment at great location"
              className="property__image"
              src="img/apartment-01.jpg"
            />
          </div>
        </div>
      </div>
      <div
        className="property__container container"
      >
        <div
          className="property__wrapper"
        >
          <div
            className="property__mark"
          >
            <span>
              Premium
            </span>
          </div>
          <div
            className="property__name-wrapper"
          >
            <h1
              className="property__name"
            >
              Beautiful & luxurious apartment at great location
            </h1>
            <Memo(ConnectFunction)
              className="property"
              height={33}
              isInBookmarks={false}
              offerId={1}
              width={31}
            />
          </div>
          <div
            className="property__rating rating"
          >
            <div
              className="property__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                Rating
              </span>
            </div>
            <span
              className="property__rating-value rating__value"
            >
              4.65
            </span>
          </div>
          <ul
            className="property__features"
          >
            <li
              className="property__feature property__feature--entire"
            />
            <li
              className="property__feature property__feature--bedrooms"
            >
              2 Bedrooms
            </li>
            <li
              className="property__feature property__feature--adults"
            >
              Max 
              4 adults
            </li>
          </ul>
          <div
            className="property__price"
          >
            <b
              className="property__price-value"
            >
              €
              120
            </b>
            <span
              className="property__price-text"
            >
               night
            </span>
          </div>
          <div
            className="property__inside"
          >
            <h2
              className="property__inside-title"
            >
              What's inside
            </h2>
            <ul
              className="property__inside-list"
            >
              <li
                className="property__inside-item"
              >
                Something
              </li>
            </ul>
          </div>
          <div
            className="property__host"
          >
            <h2
              className="property__host-title"
            >
              Meet the host
            </h2>
            <div
              className="property__host-user user"
            >
              <div
                className="property__avatar-wrapper user__avatar-wrapper property__avatar-wrapper--pro"
              >
                <img
                  alt="Host avatar"
                  className="property__avatar user__avatar"
                  height="74"
                  src="/apartment-01.jpg"
                  width="74"
                />
              </div>
              <span
                className="property__user-name"
              >
                Name
              </span>
              <span
                className="property__user-status"
              >
                Pro
              </span>
            </div>
            <div
              className="property__description"
            >
              <p
                className="property__text"
              >
                description
              </p>
            </div>
          </div>
          <Memo(ConnectFunction)
            offerId={1}
          />
        </div>
      </div>
      <section
        className="property__map map"
      >
        <CitiesMap
          activeItem={
            Object {
              "bedrooms": 2,
              "cityCoords": Array [
                52.38333,
                4.9,
              ],
              "cityName": "Amsterdam",
              "cityZoom": 15,
              "coordinates": Array [
                52.3909553943508,
                4.85309666406198,
              ],
              "description": "description",
              "goods": Array [
                "Something",
              ],
              "host": Object {
                "avatar": "apartment-01.jpg",
                "id": 2,
                "isPro": true,
                "name": "Name",
              },
              "id": 1,
              "images": Array [
                "img/apartment-01.jpg",
              ],
              "isInBookmarks": false,
              "isPremium": true,
              "maxAdults": 4,
              "photoUrl": "img/apartment-01.jpg",
              "price": 120,
              "rating": 4.65,
              "title": "Beautiful & luxurious apartment at great location",
              "type": "Apartment",
              "zoom": 16,
            }
          }
          apartmentsCoords={
            Array [
              Object {
                "coordinates": Array [
                  52.3809553943508,
                  4.939309666406198,
                ],
                "id": 4,
                "zoom": 16,
              },
              Object {
                "coordinates": Array [
                  52.3909553943508,
                  4.85309666406198,
                ],
                "id": 1,
                "zoom": 16,
              },
            ]
          }
          city={
            Object {
              "coords": Array [
                52.38333,
                4.9,
              ],
              "name": "Amsterdam",
              "zoom": 15,
            }
          }
        />
      </section>
    </section>
    <div
      className="container"
    >
      <section
        className="near-places places"
      >
        <h2
          className="near-places__title"
        >
          Other places in the neighbourhood
        </h2>
        <CardList
          apartments={
            Array [
              Object {
                "bedrooms": 2,
                "cityCoords": Array [
                  52.38333,
                  4.9,
                ],
                "cityName": "Amsterdam",
                "cityZoom": 15,
                "coordinates": Array [
                  52.3809553943508,
                  4.939309666406198,
                ],
                "description": "description",
                "goods": Array [
                  "Something",
                ],
                "host": Object {
                  "avatar": "apartment-01.jpg",
                  "id": 2,
                  "isPro": true,
                  "name": "Name",
                },
                "id": 4,
                "images": Array [
                  "img/apartment-01.jpg",
                ],
                "isInBookmarks": false,
                "isPremium": false,
                "maxAdults": 4,
                "photoUrl": "img/apartment-03.jpg",
                "price": 180,
                "rating": 4,
                "title": "Nice, cozy, warm big bed apartment",
                "type": "Apartment",
                "zoom": 16,
              },
            ]
          }
          cardClass="near-places__card"
          cardListClass="near-places__list"
        />
      </section>
    </div>
  </main>
</React.Fragment>
`;
